### ISSUE TITLE
AttributeError: 'Tensor' object has no attribute 'log_prob'

### ISSUE THREAD

**Issue by nbro** [tensorflow/tensorflow#36181]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Mac OS Catalina (Version: 10.15.2 (19C57))
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): 2.1
- Python version: 3.7.5
- GPU model and memory: Intel Iris Pro 1536 MB

**Describe the current behavior**

I get the error

> AttributeError: 'Tensor' object has no attribute 'log_prob'

with TensorFlow Probability 0.9 (and TF 2.1).

**Describe the expected behavior**

No error.

**Code to reproduce the issue**

The following code

```
import tensorflow as tf
import tensorflow_probability as tfp
from tensorflow_probability import distributions as tfd


def get_mnist_data(normalize=True):
    img_rows, img_cols = 28, 28
    (x_train, y_train), (x_test, y_test) = tf.keras.datasets.mnist.load_data()

    if tf.keras.backend.image_data_format() == 'channels_first':
        x_train = x_train.reshape(x_train.shape[0], 1, img_rows, img_cols)
        x_test = x_test.reshape(x_test.shape[0], 1, img_rows, img_cols)
        input_shape = (1, img_rows, img_cols)
    else:
        x_train = x_train.reshape(x_train.shape[0], img_rows, img_cols, 1)
        x_test = x_test.reshape(x_test.shape[0], img_rows, img_cols, 1)
        input_shape = (img_rows, img_cols, 1)

    x_train = x_train.astype('float32')
    x_test = x_test.astype('float32')

    if normalize:
        x_train /= 255
        x_test /= 255

    return x_train, y_train, x_test, y_test, input_shape


def get_bayesian_cnn(input_shape, num_classes=10):
    model_input = tf.keras.layers.Input(shape=input_shape)

    # kernel_divergence_fn=None to solve a symbolic exception.
    x = tfp.layers.Convolution2DFlipout(6, kernel_size=(5, 5), padding="SAME", activation=tf.nn.relu,
                                        kernel_divergence_fn=None)(model_input)
    x = tf.keras.layers.Flatten()(x)
    x = tfp.layers.DenseFlipout(84, activation=tf.nn.relu)(x)
    x = tfp.layers.DenseFlipout(num_classes)(x)

    model_output = tfp.layers.DistributionLambda(lambda t: tfd.Categorical(logits=t, validate_args=True))(x)

    model = tf.keras.Model(model_input, model_output)

    return model


def neg_log_likelihood(y_true, y_pred):
    return -tf.reduce_mean(y_pred.log_prob(tf.cast(tf.argmax(y_true, axis=-1), tf.int32)))


def train():
    x_train, y_train, x_test, y_test, input_shape = get_mnist_data()

    model = get_bayesian_cnn(input_shape=input_shape)

    model.compile(optimizer=tf.keras.optimizers.Adam(), loss=neg_log_likelihood,
                  metrics=[neg_log_likelihood])

    model.fit(x_train, y_train, batch_size=128, epochs=1, verbose=1)


if __name__ == "__main__":
    train()
```

**Comments**

This error seems to be due to the fact that `y_pred` is a tensor when the loss is called, while it should be a distribution. Meanwhile, I found a [question on Stack Overflow related to the third issue I mentioned above](https://stackoverflow.com/q/59743872/3924118). 

(_This is a duplicate issue of https://github.com/tensorflow/probability/issues/742, but, for completeness, I decided to open it here too._)

**Comment by gadagashwini** [comment]
Could able to reproduce the issue with Tf 2.1.
Please take a look at the gist [here](https://colab.research.google.com/gist/gadagashwini/a64c7d6f31212ac2b3b0653989d45b32/untitled359.ipynb). Thanks!

**Comment by nbro** [comment]
See https://github.com/keras-team/keras/issues/4506#issue-191721607.

**Comment by jvishnuvardhan** [comment]
@nbro I think the error is coming from the line `model_output` while creating the model. 

`model_output = tfp.layers.DistributionLambda(lambda t: tfd.Categorical(logits=t[0], validate_args=True))(x)
`

The shape of output layer is ((), ()) while the metric tries to access the `index -1` and throws the following error 

`InvalidArgumentError: slice index -1 of dimension 0 out of bounds. for 'metrics_9/neg_log_likelihood/strided_slice' (op: 'StridedSlice') with input shapes: [0], [1], [1], [1] and with computed input tensors: input[1] = <-1>, input[2] = <0>, input[3] = <1>.`

Please check the model summary below.

```
Model: "model_12"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
=================================================================
input_13 (InputLayer)        [(None, 28, 28, 1)]       0         
_________________________________________________________________
conv2d_flipout_12 (Conv2DFli (None, 28, 28, 6)         306       
_________________________________________________________________
flatten_12 (Flatten)         (None, 4704)              0         
_________________________________________________________________
dense_flipout_24 (DenseFlipo (None, 84)                790356    
_________________________________________________________________
dense_flipout_25 (DenseFlipo (None, 10)                1690      
_________________________________________________________________
distribution_lambda_11 (Dist ((), ())                  0         
=================================================================
Total params: 792,352
Trainable params: 792,352
Non-trainable params: 0
```

**Comment by nbro** [comment]
@jvishnuvardhan Yes, you're right. The problem is `t[0]`. If you pass `t` to logits, the output of the NN will now be `distribution_lambda (Distrib ((None,), (None,))` and we don't get the `InvalidArgumentError` error anymore. However, we still get the other error that I describe in the duplicate issue in the TFP issue tracker

> AttributeError: 'Tensor' object has no attribute 'log_prob'

Another user on Stack Overflow also reported this issue: https://stackoverflow.com/q/59743872/3924118.

**Comment by jvishnuvardhan** [comment]
@nbro It is clearly showing the root-cause of the error. When I added `print(help(y_pred))` in the loss function, it  showed `Tensor in module tensorflow.python.framework.ops object`. As a tensor object, it doesn't have any `log_prob` attribute. So, it is throwing `AttributeError`. Thanks!

**Comment by jvishnuvardhan** [comment]
I think this was resolved. I am closing the issue. Please feel free to open if the issue persists again. Thanks!

**Comment by nbro** [comment]
@jvishnuvardhan The issue https://github.com/tensorflow/probability/issues/742 (which is now the title of this issue) was not resolved yet, so you should not have closed this issue.

**Comment by jvishnuvardhan** [comment]
@nbro I think this is more related to `tensorflow/probability` and you already opened issue there. Anyway I am opening this here. Thanks!

**Comment by ShikhaSingh10** [comment]
Do we have solution to this  issue?

**Comment by Strateus** [comment]
Need it too

**Comment by Saduf2019** [comment]
I am able to replicate this issue on tf-nightly (2.4.0-dev20200929), please find the [gist here](https://colab.research.google.com/gist/Saduf2019/2a7de31fddf7d8b5d942bfd7aefd9513/untitled418.ipynb).

**Comment by sachinprasadhs** [comment]
Was able to reproduce your issue in Tf Nightly 2.6.0-dev20210524, please find the gist [here](https://colab.research.google.com/gist/sachinprasadhs/a32575d4f526d8508e2820bd804ab485/35650.ipynb). Thanks!

**Comment by AndiBerber** [comment]
The problem still occurs with tensorflow=2.15 and ensorflow-probability=0.23. None of the approaches in `https://github.com/tensorflow/probability/issues/742` works for me. Exporting for inference works, but the training signature throws with AttributeError: 'SymbolicTensor' object has no attribute 'log_prob'

**Comment by Venkat6871** [comment]
Hi @nbro ,
Thank you so much for filing this docs-feature request! We truly appreciate your support. At the moment, this request is not prioritized and mark it for community contributions. If you would like to contribute on a fix, please feel free to fork the repository and submit a pull request in accordance with the TensorFlow [CONTRIBUTION GUIDELINES](https://www.tensorflow.org/community/contribute/code) and link this issue in your PR to ensure the maintainers can easily review and collaborate.To avoid duplicate work, please leave a comment here if you decide to take it on!
Thank you again for your interest and we look forward to seeing your contributions!
